File: server/src/logger.cpp
#include "logger.h"
#include <iostream>
#include <sstream>
#include <iomanip>

Logger::Logger() : consoleOutput(true) {}

Logger::~Logger() {
    if (logFile.is_open()) {
        logFile.close();
    }
}

Logger& Logger::getInstance() {
    static Logger instance;
    return instance;
}

void Logger::log(Level level, const std::string& message) {
    std::lock_guard<std::mutex> lock(mutex);

    std::string timestamp = getCurrentTime();
    std::string levelStr = levelToString(level);
    std::string fullMessage = "[" + timestamp + "] [" + levelStr + "] " + message;

    if (consoleOutput) {
        std::cout << fullMessage << std::endl;
    }

    if (logFile.is_open()) {
        logFile << fullMessage << std::endl;
        logFile.flush();
    }
}

void Logger::setLogFile(const std::string& filename) {
    std::lock_guard<std::mutex> lock(mutex);
    if (logFile.is_open()) {
        logFile.close();
    }
    logFile.open(filename, std::ios::app);
    if (!logFile.is_open()) {
        std::cerr << "Unable to open log file: " << filename << std::endl;
    }
}

void Logger::setConsoleOutput(bool enabled) {
    consoleOutput = enabled;
}

std::string Logger::getCurrentTime() {
    time_t now = time(nullptr);
    struct tm* timeinfo = localtime(&now);

    std::ostringstream oss;
    oss << std::put_time(timeinfo, "%Y-%m-%d %H:%M:%S");
    return oss.str();
}

std::string Logger::levelToString(Level level) {
    switch (level) {
        case INFO: return "INFO";
        case WARNING: return "WARNING";
        case ERROR: return "ERROR";
        default: return "UNKNOWN";
    }
}

File: client/src/main/java/cz/zcu/kiv/ups/sp/Logger.java
package cz.zcu.kiv.ups.sp;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.text.SimpleDateFormat;
import java.util.Date;

public class Logger {

    public enum Level {
        INFO,
        WARNING,
        ERROR
    }

    private static final Logger instance = new Logger();

    private PrintWriter logFile;
    private Logger() {
    }

    public static Logger getInstance() {
        return instance;
    }

    public void log(Level level, String message) {
        String timestamp = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());
        String fullMessage = "[" + timestamp + "] [" + level.name() + "] " + message;

        if (level == Level.ERROR) {
            System.err.println(fullMessage);
        } else {
            System.out.println(fullMessage);
        }

        if (logFile != null) {
            logFile.println(fullMessage);
            logFile.flush();
        }
    }

    public void setLogFile(String filename) {
        try {
            if (logFile != null) {
                logFile.close();
            }
            logFile = new PrintWriter(new FileWriter(filename, false));
        } catch (IOException e) {
            log(Level.ERROR,"Unable to open log file: " + filename);
        }
    }

    public static void info(String message) {
        getInstance().log(Level.INFO, message);
    }

    public static void warning(String message) {
        getInstance().log(Level.WARNING, message);
    }

    public static void error(String message) {
        getInstance().log(Level.ERROR, message);
    }
}
